docs/static/includes/specification/int.spec.nest.nested.tf

```diff
@@ -0,0 +1,46 @@
+# The resource has nested dynamic blocks
+
+# Variable declaration
+variable "optional_multi_block" {
+  type = map(object({
+    name         = string
+    length       = optional(number)
+    nested_block = optional(object(any))
+    nested_multi_blocks = optional(map(object({
+      name   = string
+      length = optional(number)
+    })))
+  }))
+  default = null

```

Comment:
For collection values we recommend default is empty map/list/set and set nullable to false

Reply:
Hey @matt-FFFFFF, thank you for the feedback. Why is this the recommendation?

Reply:
Hi @MariusStorhaug,

We prefer this because then the author does not need to check for a null value before using the variable in an expression. It is either empty, or it is not.

The only time we recommend otherwise is if there is a semantic difference between null and an empty collection.

In most resource modules that are authoritative about the resources they deploy, nullable = false is the correct approach.

Reply:
I was instead thinking of optimizing for the 'pattern creator' or a direct consumer of the resource module. To me it makes more sense having checks in the code and have a clear interface where we support both `null` and empty (`[]`, `{}`). That would create a clear definition that `null` = don't use the feature, empty block (`[]`, `{}`) = use the default values in `optional`. This is usually what would happen if you create variables that would map to the modules that a pattern would consume and you want full flexibility to the consumed resource module.

Reply:
As I said before this very much depends on whether there is a semantic meaning to `null` vs `{}`.

In the case of a map being used in a for_each expression, there is not. Therefore it makes sense to simplify the authoring experience and set nullable to false.

In the case where we are defining an absolute list of something, and the difference between `null` and `{}` would mean that supplying an empty collection would affect existing resources, then of course we should make a sensible decision. However in this case `map()` may not be the correct type to use.

I still believe that for 80% of the cases that `nullable = false` is correct for most top level collection values. Of course there are always exceptions.